docs/birthday-kata.md

```diff
@@ -0,0 +1,890 @@
+      but I expected:  FirstName
+  
+      9 | unique type LastName = LastName Text
+     10 | unique type Name = 
+     11 |   { first: FirstName
+      .
+     23 |     employee = Employee id (Name lastName firstName) (Date 1 1 1991)
+```
+Of course, you can still make a mistake like
+```unison
+firstName = LastName "Lennon"
+```
+But it's much harder to make this mistake. If we are ok with the extra types and the overhead of deconstructing them, we can catch many more typos/bugs at compile time, which means fewer tests are necessary to ensure correctness!
+
+2) In the future, Unison may introduce the ability to make to make data constructors private and/or refined types. Doing so would allow us to to avoid another huge source of bugs when using primitive types directly, which is that the set of possible values is much larger than what we are attempting to model. You can see this in our model for `Date`
+
+```unison
+unique type Date = 
+  { day: Nat
+  , month: Nat
+  , year: Nat
+  }
+```
+
+By modelling day/month/year all as `Nat` we have introduced the possibility of an illegal state into our system. For example, in this model, we could have illegal dates like `0/500/99999999999999`. In fact, lets try to restrict the possible values a bit, to reduce the surface area for future bugs. This means more tests and more runtime validation (and runtime validation failures).
+
+Lets try to contrain the possible values a bit.
+```unison
+unique type Date = 
+  { day: Day
+  , month: Month
+  , year: Year
+  }
+unique type Month 
+  = January
+  | February
+  | March
+  | April
+  | May
+  | June
+  | July
+  | August
+  | September
+  | October
+  | November
+  | December
+unique type Day = Day Nat
+unique type Year = Year Nat
+```
+
+It is really easy for us to just enumerate all of the possible months. Now we have eliminated the possiblity of an invalid month appearing in our system! Now we have to decide how far we want to go to make illegal state unrepresentable. Maybe `Nat` is ok for year if we don't care about `BCE` and we are ok with possibly talking about birthdays millions of years in the future (not ideal). But it would be nice to disallow days of months that don't exist, and this is where refined types, dependent types, or private constructors could be leveraged. Hopefully, we will have those one day!
+
+### Email
+Let's apply some similar decision making to get an Email.
+```unison
+unique type Email = 
+  { to: EmailAddress
+  , subject: Subject
+  , body: Body
+  }
+
+unique type EmailAddress = EmailAddress Text
+unique type Subject = Subject Text
+unique type Body = Body Text
+```
+
+Let's also update `Employee` to have an email address
+```unison
+unique type Employee  = 
+  { id: EmployeeId
+  , name: Name
+  , birthday: Date 
+  , email: EmailAddress
+  }
+```
+
+It would be nice to add some constraints to the email address (like a regex to ensure only valid email addresses can exist) but this is pretty good so far!
```

Comment:
I think we could remove this sentence because in all likelihood we wouldn't use a regex to do validation of emails, and we probably couldn't solely use types to delimit the full email rfc. You've made the point really well earlier that types can make illegal states impossible so the details of how we'd validate emails can probably be hand-waved away for now 😄.

Reply:
Done but it would also be cool to get refined/dependent/ liquid types ;) And also maybe access modifiers (if we are content with run time validation)